lib/zbc_ata.c

```diff
@@ -569,33 +555,102 @@ static ssize_t zbc_ata_native_pread(struct zbc_device *dev, void *buf,
 	 * | 15  |                           Control                                     |
 	 * +=============================================================================+
 	 */
-	cmd.io_hdr.dxfer_direction = SG_DXFER_FROM_DEV;
-	cmd.cdb[0] = ZBC_SG_ATA16_CDB_OPCODE;
+	cmd->io_hdr.dxfer_direction = dxfer_dir;
+	cmd->cdb[0] = ZBC_SG_ATA16_CDB_OPCODE;
 	/* DMA protocol, ext=1 */
-	cmd.cdb[1] = (0x6 << 1) | 0x01;
+	cmd->cdb[1] = (0x6 << 1) | 0x01;
 	/* off_line=0, ck_cond=0, t_type=0, t_dir=1, byt_blk=1, t_length=10 */
-	cmd.cdb[2] = 0x0e;
-	cmd.cdb[5] = (lba_count >> 8) & 0xff;
-	cmd.cdb[6] = lba_count & 0xff;
-	cmd.cdb[7] = (lba_offset >> 24) & 0xff;
-	cmd.cdb[8] = lba_offset & 0xff;
-	cmd.cdb[9] = (lba_offset >> 32) & 0xff;
-	cmd.cdb[10] = (lba_offset >> 8) & 0xff;
-	cmd.cdb[11] = (lba_offset >> 40) & 0xff;
-	cmd.cdb[12] = (lba_offset >> 16) & 0xff;
-	cmd.cdb[13] = 1 << 6;
-	cmd.cdb[14] = ZBC_ATA_READ_DMA_EXT;
+	cmd->cdb[2] = 0x0e;
+	cmd->cdb[5] = (lba_count >> 8) & 0xff;
+	cmd->cdb[6] = lba_count & 0xff;
+	cmd->cdb[7] = (lba_offset >> 24) & 0xff;
+	cmd->cdb[8] = lba_offset & 0xff;
+	cmd->cdb[9] = (lba_offset >> 32) & 0xff;
+	cmd->cdb[10] = (lba_offset >> 8) & 0xff;
+	cmd->cdb[11] = (lba_offset >> 40) & 0xff;
+	cmd->cdb[12] = (lba_offset >> 16) & 0xff;
+	cmd->cdb[13] = 1 << 6;
+	cmd->cdb[14] = dma_ext_op;
 
 	/* Execute the command */
-	ret = zbc_sg_cmd_exec(dev, &cmd);
+	ret = zbc_sg_cmd_exec(dev, cmd);
 	if (ret != 0) {
 		/* Request sense data */
-		if (ret == -EIO && zbc_ata_sense_data_enabled(&cmd) )
+		if (ret == -EIO && zbc_ata_sense_data_enabled(cmd) )
 			zbc_ata_request_sense_data_ext(dev);
 	} else {
-		ret = (sz - cmd.io_hdr.resid) >> 9;
+		ret = (sz - cmd->io_hdr.resid) >> 9;
+	}
+
+	return ret;
+}
+
+/**
+ * Read from a ZAC device using READ DMA EXT packed
+ * in an ATA PASSTHROUGH command.
+ */
+static ssize_t zbc_ata_native_pread(struct zbc_device *dev, void *buf,
+				    size_t count, uint64_t offset)
```

Comment:
I would completely remove this one and simply handle regular pread(buf, count, offset) with preadv(&iov, 1, offset) (a regular pread() is a preadv() with one vec. That removes code. Same comment applies to all drivers, so basically to the drv ops definitions.

Reply:
That sounds good, but should the same be done for sg driver? Because with sg you do not get the direct I/O behavior if you funnel all requests over the vector I/O interface.

Reply:
I am almost done with modifying your PR to remove all the pread()/pwrite() interface and process everything using preadv()/pwritev(). Similarly to before, there is only a single path for all reads/write calls. Using your optimization for the special but most common case of iovcnt==1, direct I/Os are preserved as is for that case. Will post soon in staging branch after some testing.